ldms/src/sampler/zfs_topvdevs/zfs_topvdevs.c

```diff
@@ -0,0 +1,440 @@
+        base_set_delete(sampler_base);
+
+        new_heap_size  = previous_heap_size;
+        new_heap_size += zpool_vdev_heap_sz;
+
+        if (base_set_new_heap(sampler_base, new_heap_size) == NULL) {
+            rc = errno;
+           ldmsd_log(LDMSD_LERROR,
+                     SAMP" : Failed to resize metric set heap: %d\n", errno);
+        } else {
+		log_fn(LDMSD_LDEBUG, "ldms resize of list successful\n");
+	}
+        return rc;
+}
+
+
+static int sample(struct ldmsd_sampler *self)
+{
+        int rc = 0;
+
+        base_sample_begin(sampler_base);
+
+	list_handle = ldms_metric_get(sampler_base->set, index_store.vdev_list_idx);
+        ldms_list_purge(sampler_base->set, list_handle);
+
+        rc = zpool_iter(g_zfs, get_stats, NULL);
+        if (rc != 0) {
+            log_fn(LDMSD_LERROR, SAMP" sample():zfs_pool print_stat() failed: %d\n", rc);
+            base_sample_end(sampler_base);
+            goto err1;
+        }
+	/* this is where the rubber meets the pavement */
+
+        base_sample_end(sampler_base);
+
+err1:
+        return rc;
+}
+
+static void term(struct ldmsd_plugin *self)
+{
+        log_fn(LDMSD_LDEBUG, SAMP" term() called\n");
+        base_set_delete(sampler_base);
+        base_del(sampler_base);
+        sampler_base = NULL;
+}
+
+static ldms_set_t get_set(struct ldmsd_sampler *self)
+{
+	return NULL;
+}
+
+static const char *usage(struct ldmsd_plugin *self)
+{
+        log_fn(LDMSD_LDEBUG, SAMP" usage() called\n");
+	return  "config name=" SAMP " " BASE_CONFIG_SYNOPSIS
+                BASE_CONFIG_DESC
+                ;
+}
+
+struct ldmsd_plugin *get_plugin(ldmsd_msg_log_f pf)
+{
+        static struct ldmsd_sampler plugin = {
+                .base = {
+                        .name = SAMP,
+                        .type = LDMSD_PLUGIN_SAMPLER,
+                        .term = term,
+                        .config = config,
+                        .usage = usage,
+                },
+                .get_set = get_set,
+                .sample = sample,
+        };
+
+        log_fn = pf;
+        log_fn(LDMSD_LDEBUG, SAMP" get_plugin() called ("PACKAGE_STRING")\n");
+
+        return &plugin.base;
+}
+
+
+/*
+ * vdev summary stats are a combination of the data shown by
+ *  zpool status` and `zpool list -v
+ *  zpoolname
+ *  state
+ *  vdevname
+ *  alloc
+ *  free
+ *  size
+ *  read_bytes
+ *  read_errors
+ *  read_ops
+ *  write_bytes
+ *  write_errors
+ *  write_ops
+ *  checksum_errors
+ *  fragmentation
+ *  init_errors
+ */
+static int get_vdev_stats(nvlist_t *nvroot, const char *pool_name,
+    zpool_handle_t *zhp)
+{
+        uint_t c;
+        vdev_stat_t *vs;
+        char *vdev_name = NULL;
+	ldms_mval_t record_instance;
+        int rc=0; /*return code*/
+
+	vdev_name = zpool_vdev_name(g_zfs, zhp, nvroot, VDEV_NAME_TYPE_ID);
+
+	if (nvlist_lookup_uint64_array(	nvroot,
+					ZPOOL_CONFIG_VDEV_STATS,
+					(uint64_t **)&vs, &c) != 0) {
+                rc = 1;
```

Comment:
I'd assign `rc` to the `nvlist_lookup_uint64_array` return value which is an errno.